File: MyMonitorHub.JavaAgent/src/main/java/mymonitorhub/agent/core/PluginThread.java
package mymonitorhub.agent.core;

import com.fasterxml.jackson.databind.JsonNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Date;

/**
 * Runs a single plugin on its hibernate schedule.
 * Plugin classes are resolved via Class.forName() — the assembly config field
 * is ignored since all plugins ship inside the same fat JAR.
 * Mirrors C# PluginThread.
 */
public class PluginThread implements IThreadShutdown {

    private static final Logger log = LoggerFactory.getLogger(PluginThread.class);
    private static final int MINIMUM_SLEEP_MS = 30000;

    private volatile boolean stopped = true;
    private final Object padlock = new Object();
    private Thread thread;
    private Date lastRun = new Date(0);

    AbstractPlugin plugin;
    private String title;
    private final boolean multi;
    private final int index;

    public PluginThread(int index, boolean multi) {
        this.index = index;
        this.multi = multi;
        init();
    }

    private void init() {
        JsonNode config = Monitor.readConfig();
        JsonNode pluginArray = multi ? config.get("group-plugins") : config.get("plugins");
        if (pluginArray == null || !pluginArray.isArray() || index >= pluginArray.size()) {
            throw new RuntimeException("Plugin index " + index + " out of range");
        }
        JsonNode element = pluginArray.get(index);
        title = element.path("title").asText("plugin-" + index);
        String className = element.path("class-name").asText(null);
        if (className == null) {
            throw new RuntimeException("Plugin entry " + index + " has no class-name");
        }

        log.info("Loading plugin class: {}", className);
        try {
            Class<?> clazz = Class.forName(className);
            Object instance = clazz.newInstance();
            if (!(instance instanceof AbstractPlugin)) {
                throw new RuntimeException("class (" + className + ") must extend AbstractPlugin");
            }
            plugin = (AbstractPlugin) instance;
            plugin.initialize(element, title);
        } catch (ClassNotFoundException e) {
            throw new RuntimeException("Plugin class not found: " + className, e);
        } catch (InstantiationException | IllegalAccessException e) {
            throw new RuntimeException("Cannot instantiate plugin class: " + className, e);
        }
    }

    private int sleepTime() {
        long hibernateMs = plugin.getHibernateSeconds() * 1000L;
        long usedMs = new Date().getTime() - lastRun.getTime();
        long sleep = hibernateMs - usedMs;
        return (int) Math.max(sleep, MINIMUM_SLEEP_MS);
    }

    public void start() {
        thread = new Thread(new Runnable() {
            @Override
            public void run() {
                doRun();
            }
        });
        thread.setName("plugin-" + title);
        thread.setDaemon(true);
        thread.start();
    }

    @Override
    public void stop() {
        stopped = true;
        synchronized (padlock) {
            padlock.notifyAll();
        }
        if (thread != null) {
            try { thread.join(10000); } catch (InterruptedException ignored) {}
        }
    }

    private void doRun() {
        log.info("starting {} thread. Hibernation time = {} seconds.",
                title, plugin.getHibernateSeconds());
        if (multi) {
            ((GroupPlugin) plugin).displayPlugins();
        }
        stopped = false;

        while (!stopped) {
            lastRun = new Date();
            try {
                plugin.start();
            } catch (Exception e) {
                log.error("Plugin execution error: {}", e.getMessage());
            }

            int sleep = sleepTime();
            log.debug("going to sleep for {} seconds", sleep / 1000);
            try {
                synchronized (padlock) {
                    padlock.wait(sleep);
                }
            } catch (InterruptedException e) {
                stopped = true;
            }
        }
        log.info("{}: {} - has come to an end",
                multi ? "MultiPluginThread" : "PluginThread", title);
    }
}

File: MyMonitorHub.Agent/MyMonitorHub.Agent.Core/Core/PluginThread.cs
using System;
using System.IO;
using System.Reflection;
using System.Threading;
using Newtonsoft.Json.Linq;
using Serilog;

namespace MyMonitorHub.Agent.Core
{
    public class PluginThread : IThreadShutdown
    {
        private static readonly ILogger Logger = Log.ForContext<PluginThread>();

        private const int MinimumSleep = 30000;
        private readonly bool _multi;
        private string? _assemblyName;
        private string? _className;
        private int _hibernateSeconds;

        private DateTime _lastRun;
        internal AbstractPlugin? Plugin;
        private string? _title;

        private volatile bool _stopped = true;
        private readonly object _padlock = new object();

        public PluginThread(int index, bool multi)
        {
            // this constructor is only used from GroupPluginThread
            _multi = multi;

            var config = Monitor.ReadConfig();

            if (multi)
            {
                var groupPluginArray = config["group-plugins"] as JArray;
                if (groupPluginArray != null && index < groupPluginArray.Count)
                {
                    Init(groupPluginArray[index]);
                }
            }
            else
            {
                var pluginArray = config["plugins"] as JArray;
                if (pluginArray != null && index < pluginArray.Count)
                {
                    Init(pluginArray[index]);
                }
            }
        }

        private int SleepTime
        {
            get
            {
                var hibernateMillis = Plugin!.HibernateSeconds*(long) 1000;

                var now = DateTime.Now;
                var usedMillis = (now - _lastRun).TotalMilliseconds;

                var sleepTime = hibernateMillis - usedMillis;
                if (sleepTime < MinimumSleep)
                {
                    sleepTime = MinimumSleep;
                }

                return (int)sleepTime;
            }
        }

        private void Init(JToken element)
        {
            _title = element["title"]?.ToString();
            _className = element["class-name"]?.ToString();
            _assemblyName = element["assembly"]?.ToString();

            if (_assemblyName != null)
            {
                Logger.Information("assemblyName is: {0}", _assemblyName);
            }

            object? o = null;
            try
            {
                Assembly assembly;
                if (_assemblyName == null)
                {
                    assembly = GetType().Assembly;
                }
                else
                {
                    var baseDir = Path.GetFullPath(AppContext.BaseDirectory);
                    var assemblyPath = Path.GetFullPath(
                        Path.IsPathRooted(_assemblyName)
                            ? _assemblyName
                            : Path.Combine(baseDir, _assemblyName));

                    if (!assemblyPath.StartsWith(baseDir, StringComparison.OrdinalIgnoreCase))
                        throw new InvalidOperationException(
                            $"Plugin path '{assemblyPath}' is outside the install directory and cannot be loaded.");

                    assembly = Assembly.LoadFrom(assemblyPath);
                }
                // create an instance of the class from the assembly
                o = assembly.CreateInstance(_className!);
            }
            catch (Exception e)
            {
                Logger.Error("Class name is: {0}", _className);
                Logger.Error(e.Message + "\r\n" + e.StackTrace);
                if (e.InnerException != null)
                {
                    Logger.Error(e.InnerException.Message + "\r\n" + e.InnerException.StackTrace);
                    if (e.InnerException.InnerException != null)
                    {
                        Logger.Error(e.InnerException.InnerException.Message + "\r\n" + e.InnerException.InnerException.StackTrace);
                    }
                }
            }

            if (!(o is AbstractPlugin))
            {
                throw new Exception($"class ({_className}) must implement AbstractPlugin");
            }
            Plugin = (AbstractPlugin)o;
            Plugin.Initialize(element, _title ?? string.Empty);
        }

        public void Run()
        {
            _hibernateSeconds = Plugin!.HibernateSeconds;
            Logger.Information("starting " + _title + " thread. Hibernation time = " + _hibernateSeconds + " seconds.");
            if (_multi)
            {
                var groupPlugin = (GroupPlugin) Plugin;
                groupPlugin.DisplayPlugins();
            }
            _stopped = false;
            while (!_stopped)
            {
                _lastRun = DateTime.Now;
                try
                {
                    Plugin.Start();
                }
                catch (Exception e)
                {
                    Logger.Error(e, "Plugin execution error");
                }
                try
                {
                    var sleepTime = SleepTime;
                    Logger.Debug("going to sleep for " + sleepTime/1000 + " seconds");
                    lock (_padlock)
                    {
                        System.Threading.Monitor.Wait(_padlock, sleepTime);
                    }
                }
                catch (Exception e)
                {
                    _stopped = true;
                    Logger.Error(e, "Plugin thread wait interrupted");
                }
            } // while
            if (_multi)
            {
                Logger.Information("MultiPluginThread: " + _title + " - has come to an end");
            }
            else
            {
                Logger.Information("PluginThread: " + _title + " - has come to an end");
            }
        }

        private Thread? _thread;
        public void Start()
        {
            _thread = new Thread(Run);
            _thread.Start();
        }
        public void Stop()
        {
            _stopped = true;
            lock (_padlock)
            {
                System.Threading.Monitor.Pulse(_padlock);
            }
            _thread?.Join(10000); // wait a max of 10 seconds to shut this down
        }
    }

    // class
}